### PR TITLE
WIP: SFTP unit tests

### PR DESCRIPTION
**Pull request by ianstalk**
Putting up an early WIP PR to get feedback. I couldn't find a PyTest mock SFTP fixture so I'm trying to just do it with pure mocks. I'm having trouble figuring out a way to simulate callback behavior; any suggestions?

### CODE REVIEW

File: flexget/components/ftp/sftp_client.py [hunk 1]
```diff
@@ -26,38 +26,230 @@
 logger = logger.bind(name='sftp_client')
 
 
-class SftpClient:
+class SftpError(Exception):
```
Comment (ianstalk) [inline flexget/components/ftp/sftp_client.py:29]: Sorry these changes are so huge. There are two reasons:

1) I made an `SftpClientBuilder` class and moved a bunch of logic out of `SftpClient` both for separation of concerns, and to allow for better dependency injection.
2) Objects declared below a given class can't be used as a type annotation in the constructor, which forced me to move SftpClient to the bottom.

**Comment by liiight** [comment]
This is way too big to fit in my brain. The only feedback I can give is please run black :)

File: flexget/components/ftp/sftp_client.py [hunk 2]
```diff
@@ -26,38 +26,230 @@
...
+
+    def get_null_handler(self) -> NodeHandler:
+        """
+        Builds a noop node handler suitable for use with pysftp.Connection.walktree
+        """
+        return partial(Handlers.null_node_handler)
+
+
+class Handlers:
+    @classmethod
```
Comment (paranoidi) [inline flexget/components/ftp/sftp_client.py:106]: Complicated use of classmethods, they are generally alternative constructors.

Comment (ianstalk) [reply flexget/components/ftp/sftp_client.py:106]: I basically want to use them like static methods, but with access to the other methods in the class. Would it be preferable to fully qualify them instead?

**Pull request closed without merging**